### PR TITLE
Users/shmittal/cross platform

### PR DESCRIPTION
**Pull request by 20shivangi**

### CODE REVIEW

**Comment by 20shivangi** [comment]
Supporting  cross platform AzurePowershell task for SPN credential based service endpoint.

File: Tasks/AzurePowerShellV4/task.json [hunk 1]
```diff
@@ -148,7 +148,13 @@
...
     "execution": {
         "PowerShell3": {
-            "target": "AzurePowerShell.ps1"
+            "target": "AzurepowerShell.ps1",
+            "platforms": [
+                "windows"
+            ]
+        },
+        "Node": {
+            "target": "AzurePowerShell.js"
```
Comment (vincentdass) [inline Tasks/AzurePowerShellV4/task.json]: please keep the file name in lowercase.

Comment (20shivangi) [reply Tasks/AzurePowerShellV4/task.json]: It is done

File: Tasks/AzurePowerShellV4/typings.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
```
[the remark this answers is not in the document]
Comment (20shivangi) [reply Tasks/AzurePowerShellV4/typings.json]: It is done

File: Tasks/AzurePowerShellV4/InitializeAz.ps1 [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+)
+
+$endpointObject =  ConvertFrom-Json  $endpoint
+$moduleName = "Az.Accounts"
+$module = Get-Module -Name $moduleName -ListAvailable | Sort-Object Version -Descending | Select-Object -First 1
+$module = Import-Module -Name $module.Path -Global -PassThru -Force
+$environmentName = $endpointObject.environment
+
+$psCredential = New-Object System.Management.Automation.PSCredential(
+                    $endpointObject.servicePrincipalClientID,
```
[the remark this answers is not in the document]
Comment (20shivangi) [reply Tasks/AzurePowerShellV4/InitializeAz.ps1]: yes

File: Tasks/AzurePowerShellV4/AzurePowerShell.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        let _vsts_input_errorActionPreference: string = tl.getInput('errorActionPreference', false) || 'Stop';
+        let _vsts_input_failOnStandardError = tl.getBoolInput('FailOnStandardError', false);
+        let targetAzurePs: string = tl.getInput('TargetAzurePs', false);
+        let customTargetAzurePs: string = tl.getInput('CustomTargetAzurePs', false);
+        let serviceName = tl.getInput('ConnectedServiceNameARM',/*required*/true);
+        let endpointt= await new AzureRMEndpoint(serviceName).getEndpoint();
+
+        // string constants
+        let otherVersion = "OtherVersion"
+        let latestVersion = "LatestVersion"
```
[the remark this answers is not in the document]
Comment (20shivangi) [reply Tasks/AzurePowerShellV4/AzurePowerShell.ts]: It is done

File: Tasks/AzurePowerShellV4/AzurePowerShell.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+        if (targetAzurePs == otherVersion) {
+            if (customTargetAzurePs != "") {
```
[the remark this answers is not in the document]
Comment (20shivangi) [reply Tasks/AzurePowerShellV4/AzurePowerShell.ts]: It is done

File: Tasks/AzurePowerShellV4/AzurePowerShell.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
+import fs = require('fs');
+import path = require('path');
+import os = require('os');
+import tl = require('vsts-task-lib/task');
+import tr = require('vsts-task-lib/toolrunner');
+import { AzureRMEndpoint } from 'azure-arm-rest/azure-arm-endpoint';
+import { AzureEndpoint } from 'azure-arm-rest/azureModels';
```
[the remark this answers is not in the document]
Comment (20shivangi) [reply Tasks/AzurePowerShellV4/AzurePowerShell.ts]: It is done

File: Tasks/AzurePowerShellV4/AzurePowerShell.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        if (scriptType.toUpperCase() == 'FILEPATH') {
+            if (!tl.stats(scriptPath).isFile() || !scriptPath.toUpperCase().match(/\.PS1$/)) {
+                throw new Error(tl.loc('JS_InvalidFilePath', scriptPath));
+            }
+        }
+
+        // Generate the script contents.
+        console.log(tl.loc('GeneratingScript'));
+        let contents: string[] = [];
+        let filePath3 = path.join(path.resolve(__dirname), 'InitializeAz.ps1');
```
[the remark this answers is not in the document]
Comment (20shivangi) [reply Tasks/AzurePowerShellV4/AzurePowerShell.ts]: It is done

File: Tasks/AzurePowerShellV4/AzurePowerShell.ts [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        contents.push(`${filePath3} -targetAzurePs  ${targetAzurePs} -serviceName ${serviceName} -endpoint '${endpoint}'`);
+        contents.push(`$ErrorActionPreference = '${input_errorActionPreference}'`);
```
[the remark this answers is not in the document]
Comment (20shivangi) [reply Tasks/AzurePowerShellV4/AzurePowerShell.ts]: It is done

File: Tasks/AzurePowerShellV4/AzurePowerShell.ts [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+            .arg(`. '${filePath.replace("'", "''")}'`);
+
+        let options = <tr.IExecOptions>{
+            errStream: process.stdout, // Direct all output to STDOUT, otherwise the output may appear out
+            outStream: process.stdout, // of order since Node buffers it's own STDOUT but not STDERR.
+            ignoreReturnCode: true
+        };
+
+
+        // Run bash.
```
[the remark this answers is not in the document]
Comment (20shivangi) [reply Tasks/AzurePowerShellV4/AzurePowerShell.ts]: It is done

File: Tasks/AzurePowerShellV4/AzurePowerShell.ts [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            default:
+                throw new Error(tl.loc('JS_InvalidErrorActionPreference', input_errorActionPreference));
+        }
+
+        let scriptType: string = tl.getInput('ScriptType', /*required*/true);
+        let scriptPath = tl.getPathInput('ScriptPath', false);
+        let scriptInline: string = tl.getInput('Inline', false);
+        let scriptArguments: string = tl.getInput('ScriptArguments', false);
+        let _vsts_input_errorActionPreference: string = tl.getInput('errorActionPreference', false) || 'Stop';
+        let _vsts_input_failOnStandardError = tl.getBoolInput('FailOnStandardError', false);
```
[the remark this answers is not in the document]
Comment (20shivangi) [reply Tasks/AzurePowerShellV4/AzurePowerShell.ts]: It is done

File: Tasks/AzurePowerShellV4/AzurePowerShell.ts [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        let targetAzurePs: string = tl.getInput('TargetAzurePs', false);
+        let customTargetAzurePs: string = tl.getInput('CustomTargetAzurePs', false);
+        let serviceName = tl.getInput('ConnectedServiceNameARM',/*required*/true);
+        let endpointt= await new AzureRMEndpoint(serviceName).getEndpoint();
```
[the remark this answers is not in the document]
Comment (20shivangi) [reply Tasks/AzurePowerShellV4/AzurePowerShell.ts]: It is done

File: Tasks/AzurePowerShellV4/task.json [hunk 12]
```diff
@@ -148,13 +148,21 @@
...
+        "Node": {
+            "target": "azurepowershell.js"
         }
     },
     "messages": {
         "InvalidScriptArguments0": "Invalid script arguments '{0}'. Line breaks are not allowed.",
         "InvalidScriptPath0": "Invalid script path '{0}'. Invalid path characters specified.",
         "InvalidAzurePsVersion": "The Azure PowerShell version '{0}' specified is not in the correct format. Please check the format. An example of correct format is 1.0.1",
-        "OverrideAzurePowerShellVersion": "The Azure PowerShell version '{0}' is not supported for Hosted Agent. Overriding it with version '{1}'. Consider changing the version selected."
+        "OverrideAzurePowerShellVersion": "The Azure PowerShell version '{0}' is not supported for Hosted Agent. Overriding it with version '{1}'. Consider changing the version selected.",
```
Comment (niadak) [inline Tasks/AzurePowerShellV4/task.json]: Is this getting used anymore?

File: Tasks/AzurePowerShellV4/InitializeAz.ps1 [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+param
+(
+    [String] [Parameter(Mandatory = $false)]
+    $targetAzurePs,
+
+    [String] [Parameter(Mandatory = $true)]
+    $serviceName,
+
+    [String] [Parameter(Mandatory = $true)]
+    $endpoint
```
Comment (niadak) [inline Tasks/AzurePowerShellV4/InitializeAz.ps1]: Mandatory input should come before optional imputs

File: Tasks/AzurePowerShellV4/InitializeAz.ps1 [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+# Clear context
+Write-Host "##[command]Clear-AzContext -Scope Process"
+$null = Clear-AzContext -Scope Process
+Write-Host "##[command]Clear-AzContext -Scope CurrentUser -Force -ErrorAction SilentlyContinue"
+$null = Clear-AzContext -Scope CurrentUser -Force -ErrorAction SilentlyContinue 
+
+if ($endpointObject.scheme -eq 'ServicePrincipal') {
+    try {
+        if ($endpointObject.authenticationType -ne 'SPNCertificate') {
```
Comment (niadak) [inline Tasks/AzurePowerShellV4/InitializeAz.ps1]: Check for -eq 'SPNCredential'

File: Tasks/AzurePowerShellV4/InitializeAz.ps1 [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+            $psCredential = New-Object System.Management.Automation.PSCredential(
+                    $endpointObject.servicePrincipalClientID,
+                    (ConvertTo-SecureString $endpointObject.servicePrincipalKey -AsPlainText -Force))
+            Write-Host "##[command]Connect-AzAccount -ServicePrincipal -Tenant $($endpointObject.tenantId) -Credential $psCredential -Environment $environmentName"
+            $null = Connect-AzAccount -ServicePrincipal -Tenant $endpointObject.tenantId `
+            -Credential $psCredential `
+            -Environment $environmentName -WarningAction SilentlyContinue
+        }
```
Comment (niadak) [inline Tasks/AzurePowerShellV4/InitializeAz.ps1]: Throw on else

File: Tasks/AzurePowerShellV4/InitializeAz.ps1 [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    }
+    catch {
+        # Provide an additional, custom, credentials-related error message. Will handle localization later
+        throw ("Only support SPN credentials cross platform for now in Azure PowerShell")
```
Comment (niadak) [inline Tasks/AzurePowerShellV4/InitializeAz.ps1]: We should show the actual error here

File: Tasks/AzurePowerShellV4/AzurePowerShell.ts [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+            ignoreReturnCode: true
+        };
+
+        // Listen for stderr.
+        let stderrFailure = false;
+        if (_vsts_input_failOnStandardError) {
+            powershell.on('stderr', (data) => {
+                stderrFailure = true;
+            });
+        }
```
Comment (niadak) [inline Tasks/AzurePowerShellV4/AzurePowerShell.ts]: We need to clear endpoint context

File: Tasks/AzurePowerShellV4/task.json (hunk 12 above)
Comment (20shivangi) [reply Tasks/AzurePowerShellV4/task.json]: It is done

File: Tasks/AzurePowerShellV4/InitializeAz.ps1 (hunk 13 above)
Comment (20shivangi) [reply Tasks/AzurePowerShellV4/InitializeAz.ps1]: It is done

File: Tasks/AzurePowerShellV4/InitializeAz.ps1 (hunk 14 above)
Comment (20shivangi) [reply Tasks/AzurePowerShellV4/InitializeAz.ps1]: It is done.

File: Tasks/AzurePowerShellV4/InitializeAz.ps1 (hunk 15 above)
Comment (20shivangi) [reply Tasks/AzurePowerShellV4/InitializeAz.ps1]: It is done

File: Tasks/AzurePowerShellV4/AzurePowerShell.ts (hunk 17 above)
Comment (20shivangi) [reply Tasks/AzurePowerShellV4/AzurePowerShell.ts]: Required only for SPN certificate, not for SPN credentials

**Pull request merged**